sol/Usdx.sol

```diff
@@ -1,48 +1,148 @@
+ *   - Locked eth can be redeemed by the original sender by burning USDX
+ *     at the originally minted price.
  *
- *   Note: Owner is able to set the Eth/USD oracle.
+ *   Note: Owner is able to set the eth/usd oracle.
  */
 contract USDX is ERC20, Ownable {
 	using SafeMath for uint256;
 	using SafeCast for int256;
 	uint8 constant FEED_DECS = 8;
-	// ETH/USD exchange rate feed.
-	AggregatorV3Interface public ethUsdPriceFeed;
+	AggregatorV3Interface public usdPriceFeed;
+	uint8 public feedDecs;
 
-	constructor (address _priceFeed) public ERC20("USDX Stablecoin", "USDX") {
+	struct account {
+		uint256 locked;
+		uint256 mint;
+	}
+	mapping (address => account) public accounts;
+
+	constructor (address _priceFeed) ERC20("USDX Stablecoin", "USDX") {
 		setFeed(_priceFeed);
 	}
 
+	// _newFeed's decimals parameter must immutably be set to 8.
 	function setFeed(address _newFeed) public onlyOwner {
 		AggregatorV3Interface newFeed = AggregatorV3Interface(_newFeed);
 		require(newFeed.decimals() == FEED_DECS);
-		ethUsdPriceFeed = newFeed;
+		usdPriceFeed = newFeed;
 	}
 
+	// Received eth is minted into usdx at the current eth/usd
+	// exchange rate.
+	// Note: msg.sender must be payable to allow redemption of
+	// deposited eth.
 	receive() external payable {
-		(,int256 rate,,,) = ethUsdPriceFeed.latestRoundData();
-		uint256 usd = msg.value.mul(rate.toUint256()).div(10**FEED_DECS); //latestRoundData uses 8 decimals
-		_mint(_msgSender(), usd);
+		(,int256 rate,,,) = usdPriceFeed.latestRoundData();
+		uint256 toMint = weiToUSDX(msg.value, rate);
+		account storage acct = accounts[msg.sender];
+		acct.locked += msg.value;
+		acct.mint += toMint;
+		_mint(msg.sender, toMint);
+	}
+
+	// Redeems usdx back into eth.  This method can only be called by
+	// senders who have previously directly sent eth into this
+	// contract.  The amount of eth redeemed is capped to the amount
+	// of eth the sender has previously sent.  The amount of eth
+	// redeemed is unrelated to the current ETH/USD price; it's based
+	// purely on the ratio of previously sent eth and previously
+	// minted usdx (ie: if 1 eth was previously received by this
+	// contract to mint 1000usdx, 500 usdx is able to be redeemed for
+	// .5 eth regardless of whether usd/eth price has decreased to
+	// 800usd/eth).  If _amount is 0, msg.sender's full usdx balance
+	// will be used to redeem eth.  Otherwise, _amount is the maximum
+	// amount of usdx used to redeem into eth.
+	function redeem(uint256 _amount) public returns (uint256) {
+		account storage acct = accounts[msg.sender];
+		require(acct.locked > 0, "nothing to redeem");
+		if (_amount == 0) {
+			_amount = acct.mint;
+		} else {
+			_amount = min(_amount, acct.mint);
+		}
+
+		_amount = min(_amount, balanceOf(msg.sender));
+		require(_amount > 0, "no usdx balance");
+		_burn(msg.sender, _amount);
+
+		uint256 unlock = acct.locked.mul(_amount).div(acct.mint);
+		if (_amount == acct.mint) {
+			delete accounts[msg.sender];
```

Comment:
Do we want to always delete the users account if they redeem all of their usdx? Not fully familiar with how deleting addresses/accounts works, but I'm just curious if this prevents a user from keeping the same address if they were to withdraw all their usdx and then later decide they want to deposit more again.

Reply:
This does not prevent the user from keeping the same address and depositing later.  From solidity docs:
> You can think of mappings as hash tables, which are virtually initialised such that every possible key exists and is mapped to a value whose byte-representation is all zeros
Deleting the account gives a very slight gas savings (because we're actually removing some data that needs to be stored), but they're logically equal (after deletion, `account[msg.sender].locked == 0`)